### PR TITLE
Add Ethereum provider plugin and rust wrapper 

### PR DESCRIPTION
**Pull request by fetsorn**
This PR adds an Ethereum provider plugin and an Ethereum wrapper in Rust.

Provider plugin API consists of `request(method, params)`, in accordance with the [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) standard. Methods in the `personal_` namespace are resolved locally with a signer, methods with `eth_` are sent to the chain.

The wrapper is close to parity with the existing JavaScript Ethereum plugin.

## Implementation notes:
eth/wei conversion and encoding of function input were implemented by @namesty. We parse the function ABI, extract parameter types and encode the values.

[block_on](https://docs.rs/futures/latest/futures/executor/fn.block_on.html) executes asynchronous functions from the `ethers-rs` crate.

I opt for lower-level implementations in place of those parts of `ethers-rs` API that require bindgen, i.e. most of [ethers::contract::Contract](https://docs.rs/ethers/latest/ethers/contract/struct.Contract.html) API, the `ethers::providers::Provider` [send_transaction](https://docs.rs/ethers/latest/ethers/providers/trait.Middleware.html#method.send_transaction) method and everything else connected to [PendingTransaction](https://docs.rs/ethers/latest/ethers/providers/struct.PendingTransaction.html), [FilterWatcher](https://docs.rs/ethers/latest/ethers/providers/struct.FilterWatcher.html) and [GasOracle](https://docs.rs/ethers/latest/ethers/etherscan/gas/struct.GasOracle.html). 
So, instead of building contract instances and invoking their methods, we parse arguments, build transactions, sign and send them as raw bytes.

## Future work:
Following features were left out of this scope and will arrive as separate PRs
 - txOverrides
 - web3provider
 - better error handling
 - EIP-1559 and gas oracles
 - waitForEvent, confirmations and other async calls that require timeouts

### CODE REVIEW

File: protocol/ethereum/provider/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  Module,
+  manifest,
+  Args_request,
+} from "./wrap";
+import { ethers, Wallet } from "ethers";
+import { PluginFactory } from "@polywrap/core-js";
+
+export interface ProviderConfig {
+  url: string
+  wallet?: Wallet
```
Comment (krisbitney) [inline protocol/ethereum/provider/src/index.ts:11]: Does this mean users need to have ethers as a dependency so they can provider a `Wallet` to the config?

**Review by krisbitney — CHANGES_REQUESTED** [review]
This is amazing! I'm so happy to see how much progress you've made. 

While the wrapper builds with `polywrap build`, it fails `cargo check`. Do you know why?

**Comment by cbrzn** [comment]
hey @fetsorn any updates on this? would love to use this wrapper in the python client :-D

File: protocol/ethereum/provider/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    args: Args_request
+  ): Promise<string> {
+    if (args.method === "personal_signDigest") {
+      if (this._wallet && args.params) {
+        return ethers.utils.joinSignature(
+          this._wallet._signingKey().signDigest(JSON.parse(args.params))
+        );
+      }
+    } else if (args.method === "personal_signTypedData") {
+      return "";
```
Comment (krisbitney) [inline protocol/ethereum/provider/src/index.ts:36]: This branch returns an empty string

File: protocol/ethereum/provider/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    } else if (args.method === "personal_address") {
+      if (this._wallet) {
+        return await this._wallet.getAddress();
+      }
+    } else if (args.method === "personal_chainId") {
+      const network = await this._provider.getNetwork();
+      return network.chainId.toString()
+    } else if (args.method === "personal_withChainId") {
+      return "";
```
Comment (krisbitney) [inline protocol/ethereum/provider/src/index.ts:45]: this branch returns an empty string

**Review by krisbitney — APPROVED** [review]

**Pull request merged**